core/vm/contracts.go

```diff
@@ -945,6 +945,75 @@ func (c *getValidator) Run(input []byte, caller common.Address, evm *EVM, gas ui
 	return addressBytes, gas, nil
 }
 
+type getValidatorBLS struct{}
+
+func (c *getValidatorBLS) RequiredGas(input []byte) uint64 {
+	return params.GetValidatorBLSGas
+}
+
+func copyBEtoLE(result []byte, offset int, uncompressedBytes []byte, offset2 int) {
+	for i := 0; i < 48; i++ {
+		result[63-i+offset] = uncompressedBytes[i+offset2]
+	}
+}
+
+// Return the validator BLS public key for the validator at given index. The public key is given in uncompressed format, 4*48 bytes.
+func (c *getValidatorBLS) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
+	gas, err := debitRequiredGas(c, input, gas)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	// input is comprised of two arguments:
+	//   index: 32 byte integer representing the index of the validator to get
+	//   blockNumber: 32 byte integer representing the block number to access
+	if len(input) < 64 {
+		return nil, gas, ErrInputLength
+	}
+
+	index := new(big.Int).SetBytes(input[0:32])
+
+	blockNumber := new(big.Int).SetBytes(input[32:64])
+	if blockNumber.Cmp(common.Big0) == 0 {
+		// Validator set for the genesis block is empty, so any index is out of bounds.
+		return nil, gas, ErrValidatorsOutOfBounds
+	}
+	if blockNumber.Cmp(evm.Context.BlockNumber) > 0 {
+		return nil, gas, ErrBlockNumberOutOfBounds
+	}
+
+	// Note: Passing empty hash as here as it is an extra expense and the hash is not actually used.
+	validators := evm.Context.Engine.GetValidators(new(big.Int).Sub(blockNumber, common.Big1), common.Hash{})
+
+	// Ensure index, which is guaranteed to be non-negative, is valid.
+	if index.Cmp(big.NewInt(int64(len(validators)))) >= 0 {
+		return nil, gas, ErrValidatorsOutOfBounds
+	}
+
+	validator := validators[index.Uint64()]
+	uncompressedBytes := validator.BLSPublicKeyUncompressed()
+	if len(uncompressedBytes) == 0 {
+		log.Warn("Uncompressed BLS key wasn't cached", "address", validator.Address())
+		validator.CacheUncompressed()
+		uncompressedBytes = validator.BLSPublicKeyUncompressed()
+	}
+	if len(uncompressedBytes) != 192 {
```

Comment:
shouldn't this check be done the moment we compute the uncompressed key??

are all the others callers left to do the validation themselves, as if `.BLSPublicKeyUncompressed()` could return an invalid version of this???

Reply:
It will return a string with length 0 if the compressed BLS public key is invalid. Of course, this should never be the case, but I think it's better to not crash with out of bounds access here.